dex/src/main/java/io/crate/data/join/HashInnerJoinBatchIterator.java

```diff
@@ -115,16 +125,35 @@ public void moveToStart() {
 
     @Override
     public boolean moveNext() {
-        // Build HashMap from left relation
+        while (buildBufferAndMatchRight() == false) {
+            if (right.allLoaded() && left.allLoaded()) {
+                // both sides are fully loaded, we're done here
+                return false;
+            } else if (right.allLoaded()) {
+                right.moveToStart();
+                activeIt = left;
+                buffer.clear();
+            } else {
+                // activeIt needs the next batch loaded
+                return false;
+            }
+        }
+
+        // match found
+        return true;
+    }
+
+    private boolean buildBufferAndMatchRight() {
         if (activeIt == left) {
             while (left.moveNext()) {
                 Object[] currentRow = left.currentElement().materialize();
                 int hash = hashBuilderForLeft.apply(left.currentElement());
                 addToBuffer(currentRow, hash);
             }
-            if (left.allLoaded()) {
+            if (buffer.size() == blockSize || left.allLoaded()) {
```

Comment:
I think this is not correct. If left.moveNext() returns false means the page is loaded and we need to switch to the right. The `buffer.size() == blockSize()` should be in the while loop above to exit earlier (as "safety net").

Reply:
I don't think we should operate under the assumption that an iterator batch size is the same as the page size. I think we should load up the buffer (even if it means loading multiple left iterator batches)